casper/src/test/scala/coop/rchain/casper/batch2/BlockValidationLogicSpec.scala

```diff
@@ -0,0 +1,150 @@
+package coop.rchain.casper.batch2
+
+import cats.syntax.all._
+import com.google.protobuf.ByteString
+import coop.rchain.casper.InvalidBlock.{ContainsExpiredDeploy, ContainsFutureDeploy}
+import coop.rchain.casper.ValidBlock.Valid
+import coop.rchain.casper.protocol.{BlockMessage, DeployData, ProcessedDeploy}
+import coop.rchain.casper.util.{ConstructDeploy, ProtoUtil}
+import coop.rchain.casper.{BlockValidationLogic, ValidatorIdentity}
+import coop.rchain.crypto.signatures.{Secp256k1, Signed}
+import coop.rchain.crypto.{PrivateKey, PublicKey}
+import coop.rchain.models.BlockVersion
+import coop.rchain.models.blockImplicits.{arbBlockMessage, getRandomBlock}
+import coop.rchain.models.syntax._
+import org.scalacheck.Arbitrary
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatestplus.scalacheck.ScalaCheckPropertyChecks
+
+class BlockValidationLogicSpec extends AnyFlatSpec with Matchers with ScalaCheckPropertyChecks {
+
+  implicit private val aBlock: Arbitrary[BlockMessage] = arbBlockMessage
+
+  "Block version validation" should "allow supported versions" in {
+    forAll { (block: BlockMessage, version: Int) =>
+      val blockWithVersion = block.copy(version = version)
+
+      // Expected one of hard-coded block versions supported by this version of RNode software
+      val expectedValid = BlockVersion.Supported.contains(version)
+      // Actual validation
+      val actualValid = BlockValidationLogic.version(blockWithVersion)
+
+      actualValid shouldBe expectedValid
+    }
+  }
+
+  "Block hash format validation" should "fail on invalid hash" in {
+    forAll { (block: BlockMessage) =>
+      val hash           = ProtoUtil.hashBlock(block)
+      val blockValidHash = block.copy(blockHash = hash)
+
+      // Test valid block hash
+      val hashValid = BlockValidationLogic.blockHash(blockValidHash)
+
+      hashValid shouldBe true
+
+      val blockInValidHash = block.copy(blockHash = ByteString.copyFromUtf8("123"))
+
+      // Test invalid block hash
+      val hashInValid = BlockValidationLogic.blockHash(blockInValidHash)
+
+      hashInValid shouldBe false
+    }
+  }
+
+  "Field format validation" should "succeed on a valid block and fail on empty fields" in {
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val genesis                 = signedBlock(privateKey, publicKey)
```

Comment:
It's a little confusing to call this block genesis.
---

Why not use generative testing here? We can test that empty fields are checked even when other fields are changing and also makes tests more unified.

Reply:
[Done](https://github.com/rchain/rchain/pull/3793/commits/e3451096b03ed4eca4ff9931f252f7621143c2e2). I just little bit confusing that test duplicates `formatOfFields` implementation.